### PR TITLE
Reduce doc.render() calls if possible

### PR DESCRIPTION
**Pull request by sotaroikeda**
The pull request is for reduce cpu usage by reducing doc.render() calls if possible. It is especially for reduce cpu usage during video playback [Bug 1478566](https://bugzilla.mozilla.org/show_bug.cgi?id=1478566)

It also have an effect of improving talos performance in general.

https://treeherder.mozilla.org/perf.html#/compare?originalProject=try&originalRevision=00e262c79f6716ea4815ca83749da1b17d8e56de&newProject=try&newRevision=7cb9aff96b7b2a573e8f6724d86737344200adb4&framework=1

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2994)

### CODE REVIEW

File: webrender/src/render_backend.rs [hunk 1]
```diff
@@ -702,6 +718,13 @@ impl RenderBackend {
         self.last_scene_id
     }
 
+    pub fn set_render_on_generate_frame(&mut self) {
```
Comment (NiLSPACE) [inline webrender/src/render_backend.rs]: As a complete outsider this seems confusing to me. If a method starts with `set` I assume I can change it to what I specify, but it always sets it to true. Wouldn't `enable_render_on_generate_frame` make more sense?

**Comment by gw3583** [comment]
@nical Could you take a quick look over this?

**Comment by zptan** [comment]
Sotaro will take PTO in 31/8 ~ 7/9...

File: webrender/src/render_backend.rs [hunk 2]
```diff
@@ -204,10 +207,15 @@ impl Document {
...
 
+                let scrolled = self.scroll_nearest_scrolling_ancestor(delta, node_index);
+
                 let should_render =
                     should_render &&
-                    self.scroll_nearest_scrolling_ancestor(delta, node_index) &&
+                    scrolled &&
                     self.render_on_scroll == Some(true);
+
+                self.render_on_generate_frame |= scrolled && !should_render;
```
Comment (kvark) [inline webrender/src/render_backend.rs]: could you explain the logic here please?

File: webrender/src/render_backend.rs [hunk 3]
```diff
@@ -1042,6 +1078,12 @@ impl RenderBackend {
             return;
         }
 
+        if !transaction_msg.resource_updates.is_empty() {
+            // Set render_on_generate_frame for any existing documents
+            // if resource_updates is not empty, since the updates could affects to another documents.
```
Comment (kvark) [inline webrender/src/render_backend.rs]: "could affects to another documents"

Comment (kvark) [reply webrender/src/render_backend.rs]: Why do we draw the connection between resource updates and frame generation here?

File: webrender/src/render_backend.rs (hunk 2 above)
Comment (mattwoodrow) [reply webrender/src/render_backend.rs]: The comment just above this about double borrowing is probably contributing to this being more complex than it should be.

That said, I think we could rename the earlier usage of should_render to found_scroll_node, and then make this code look like:

let mut should_render = false;
if (found_scroll_node && self.scroll_nearest_scrolling_ancestor(delta, node_index) {
  if (self.render_on_scroll == Some(true)) {
    should_render = true;
  } else {
    self.render_on_generate_frame = true;
  }
}

File: webrender/src/render_backend.rs (hunk 3 above)
Comment (mattwoodrow) [reply webrender/src/render_backend.rs]: It's not changing whether we generate a frame, it's just changing if we call render() at that time.

Before this patch we unconditionally called render() before generating a frame, now we only do so if something relevant has changed that would a new render. I assume that a resource update is one of those things, but not 100% sure.

**Comment by mattwoodrow** [comment]
> I feel a bit uneasy about the increased complexity of the internal state. Could you go into a bit more detail on what use case this change is meant to optimize and why?

Currently when playing video, we get two render() calls. Once when we update the video frame, and again during generate frame.

This patch makes the second call to render() conditional, and only does it if we actually need (we didn't render earlier, or relevant changes have been made since).

File: webrender/src/render_backend.rs (hunk 2 above)
Comment (kvark) [reply webrender/src/render_backend.rs]: (heads up - I edited your code a bit)
the suggestion sounds good to me

File: webrender/src/render_backend.rs (hunk 3 above)
Comment (kvark) [reply webrender/src/render_backend.rs]: do you mean "called render() *after* generating a frame"?

**Comment by kvark** [comment]
@mattwoodrow 

>  Once when we update the video frame, and again during generate frame.

If all Gecko needs to do when playing a video is to update the contents, why do we even go the "generate frame" route?

**Comment by kvark** [comment]
@mattwoodrow I agree, my confusion comes from this sentence:
> Once when we update the video frame, and again during generate frame.
> This patch makes the second call to render() conditional, and only does it if we actually need (we didn't render earlier, or relevant changes have been made since).

Since V-sync triggered render makes more sense to leave untouched, it should be the *first* call to render() that is conditional (currently triggered by the image update, if I understand correctly), not the second.

**Comment by mattwoodrow** [comment]
> Since V-sync triggered render makes more sense to leave untouched, it should be the first call to render() that is conditional (currently triggered by the image update, if I understand correctly), not the second.

We might have to wait for Sotaro to get back, I am running out of answers :)

My reading of the patch is that it's doing the opposite, doing a render() on the image update, and skipping the one on vsync/generate-frame.

**Comment by nical** [comment]
We need to do the render at vsync time and skip the other one, otherwise the frame is rendered with the previous animated properties which we sample and send with the vsync transaction.

**Comment by nical** [comment]
I've been experimenting with building frames only when strictly necessary. I think that we should be able to figure out the minimum amount of work needed by looking at what's in the transaction in the render backend.

One tricky aspect is that scrolling depends on having valid hit testing information while also invalidating the frame.
Since building the hit testing information happens while building the frame, we end up with a

FrameBuild <- HitTest <- FrameBuild dependny chain that breaks when we try to remove frame builds when no rendering is requested.
To fix this we can make store a flag that says whether the hit test info is up to date, and build the frame before (to get hit test info) and after (to render with new scroll offsets) handling the scroll message. This is a bit silly but technically not worse than what we have now, and the frame build happening before can be replaced with a simpler version that only builds the hit testing info as a followup optimization.

**Comment by kvark** [comment]
@nical what would helpful here to resolve the dependency roundrips is to not do a full frame rebuild in order to get the hit test info, and instead do some minimal processing there. This is #2807

**Comment by sotaroikeda** [comment]
>Since V-sync triggered render makes more sense to leave untouched, it should be the first call to render() that is conditional (currently triggered by the image update, if I understand correctly), not the second.

We need to update hittest on document render when display list is updated. And by [Bug 1478566 Comment 9](https://bugzilla.mozilla.org/show_bug.cgi?id=1478566#c9), render document during display update resulted in better scroll performance.

**Comment by sotaroikeda** [comment]
> Currently when playing video, we get two render() calls. Once when we update the video frame, and again during generate frame.

On windows, we do not get  render() for updating the video frame by  [Bug 1477608](https://bugzilla.mozilla.org/show_bug.cgi?id=1477608) if video frame format is same.

**Comment by sotaroikeda** [comment]
> On windows, we do not get render() for updating the video frame by Bug 1477608 if video frame format is same.

Then the content does not have apz change nor animation nor display list update, we do not need to call  render() during generating frame.

File: webrender/src/render_backend.rs (hunk 2 above)
Comment (sotaroikeda) [reply webrender/src/render_backend.rs]: Thanks @mattwoodrow, it is what I intended!

File: webrender/src/render_backend.rs (hunk 1 above)
Comment (sotaroikeda) [reply webrender/src/render_backend.rs]: Thanks for the advice:) Yes, it makes more sense!

File: webrender/src/render_backend.rs (hunk 3 above)
Comment (sotaroikeda) [reply webrender/src/render_backend.rs]: On windows, we do not get render() for updating the video frame by [Bug 1477608](https://bugzilla.mozilla.org/show_bug.cgi?id=1477608) if video frame format is same. Then the content does not have apz change nor animation nor display list update, we do not need to call render() during generating frame.

**Comment by sotaroikeda** [comment]
> Then the content does not have apz change nor animation nor display list update, we do not need to call render() during generating frame.

So for me  #2807 a bit orthogonal problem. This pull request is for skip render()  during generating frame if possible.

**Comment by kvark** [comment]
> We need to update hittest on document render when display list is updated

Hittest metadata shouldn't need to be changed if we just replaced the external ID of an existing image item to a new one: dimensions, formats, and positions all stay the same. This is what the "first" frame-build (of the 2) is about, if I understand correctly.

> So for me, #2807 an orthogonal problem. This pull request is for skip/suppress render() during generating frame if possible.

If we don't need to build a frame for updating hit-testing, there are no longer going to be 2 calls to render(). This is not a blocker for the PR, but I find it highly related.

**Comment by sotaroikeda** [comment]
> I have something like this partially implemented, I'll put a PR up soon-ish.

Thanks!

**Comment by sotaroikeda** [comment]
This was replaced by #3043.

**Pull request closed without merging**